File: simulated-noise-sha256-random-bits-generator.py
#Created by Tsubasa Kato with the the help of ChatGPT (GPT-4)
#(C)Tsubasa Kato - 5/8/2024 - Intended for Experimental Use.
import os
import hashlib
import numpy as np

class EntropySource:
    def __init__(self, source_size=64):
        self.source_size = source_size
        self.raw_entropy = None

    def collect_noise(self):
        """Simulate noise source collection using OS entropy."""
        self.raw_entropy = os.urandom(self.source_size)
        return self.raw_entropy

    def health_tests(self, noise):
        """Perform simple health tests on the collected noise."""
        if len(set(noise)) < len(noise) / 2:
            raise ValueError("Repetition Count Test Failed")

    def estimate_entropy(self, noise):
        """Estimate entropy using the 'min-entropy' approach."""
        counts = np.unique(list(noise), return_counts=True)[1]
        probabilities = counts / sum(counts)
        min_entropy = -np.log2(max(probabilities))
        return min_entropy

class NistSp80090B:
    def __init__(self):
        self.entropy_source = EntropySource()

    def condition(self, noise):
        """Condition the raw noise using SHA-256."""
        hash_result = hashlib.sha256(noise).digest()
        return hash_result

    def generate_random_bits(self, bit_length=256):
        """Generate random bits following NIST SP 800-90B guidelines."""
        # 1. Collect raw noise
        noise = self.entropy_source.collect_noise()

        # 2. Perform health tests
        try:
            self.entropy_source.health_tests(noise)
        except ValueError as e:
            raise RuntimeError("Health test failed") from e

        # 3. Estimate entropy
        estimated_entropy = self.entropy_source.estimate_entropy(noise)
        if estimated_entropy < 2:  # Arbitrary entropy threshold
            raise RuntimeError("Insufficient entropy")

        # 4. Condition the noise
        conditioned_output = self.condition(noise)

        # 5. Return the required number of random bits
        return conditioned_output[:bit_length // 16]

# Example usage
nist_rng = NistSp80090B()
random_bits = nist_rng.generate_random_bits()
print(random_bits.hex())


a = 1
while a <= 10000:
    nist_rng = NistSp80090B()
    random_bits = nist_rng.generate_random_bits()
    print("Number:" + str(a))
    print(random_bits.hex())
    a = a + 1 

File: simulated-noise-sha256-random-bits-generator.c
/*
Created by Tsubasa Kato with the help of ChatGPT (GPT-4)
(C) Tsubasa Kato - 5/8/2024 - Intended for Experimental Use.
Tested to work on MacOS Sonoma 14 with OpenSSL 3. (5/23/2024 - 21:51PM)
Read Readme for details on how to compile.
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <math.h>
#include <openssl/evp.h>
#include <openssl/rand.h>

#define SOURCE_SIZE 64

// Function to collect noise using OS entropy
void collect_noise(uint8_t *noise) {
    if (RAND_bytes(noise, SOURCE_SIZE) != 1) {
        fprintf(stderr, "Failed to collect noise using RAND_bytes\n");
        exit(EXIT_FAILURE);
    }
}

// Function to perform health tests on the collected noise
void health_tests(uint8_t *noise) {
    int counts[256] = {0};
    for (int i = 0; i < SOURCE_SIZE; i++) {
        counts[noise[i]]++;
    }
    int unique_count = 0;
    for (int i = 0; i < 256; i++) {
        if (counts[i] > 0) {
            unique_count++;
        }
    }
    if (unique_count < SOURCE_SIZE / 2) {
        fprintf(stderr, "Repetition Count Test Failed\n");
        exit(EXIT_FAILURE);
    }
}

// Function to estimate entropy using the 'min-entropy' approach
double estimate_entropy(uint8_t *noise) {
    int counts[256] = {0};
    for (int i = 0; i < SOURCE_SIZE; i++) {
        counts[noise[i]]++;
    }
    double probabilities[256] = {0.0};
    for (int i = 0; i < 256; i++) {
        probabilities[i] = (double)counts[i] / SOURCE_SIZE;
    }
    double max_probability = 0.0;
    for (int i = 0; i < 256; i++) {
        if (probabilities[i] > max_probability) {
            max_probability = probabilities[i];
        }
    }
    return -log2(max_probability);
}

// Function to condition the raw noise using SHA-256
void condition(uint8_t *noise, uint8_t *hash_result) {
    EVP_MD_CTX *mdctx;
    if((mdctx = EVP_MD_CTX_new()) == NULL) {
        fprintf(stderr, "EVP_MD_CTX_new failed\n");
        exit(EXIT_FAILURE);
    }

    if(1 != EVP_DigestInit_ex(mdctx, EVP_sha256(), NULL)) {
        fprintf(stderr, "EVP_DigestInit_ex failed\n");
        EVP_MD_CTX_free(mdctx);
        exit(EXIT_FAILURE);
    }

    if(1 != EVP_DigestUpdate(mdctx, noise, SOURCE_SIZE)) {
        fprintf(stderr, "EVP_DigestUpdate failed\n");
        EVP_MD_CTX_free(mdctx);
        exit(EXIT_FAILURE);
    }

    if(1 != EVP_DigestFinal_ex(mdctx, hash_result, NULL)) {
        fprintf(stderr, "EVP_DigestFinal_ex failed\n");
        EVP_MD_CTX_free(mdctx);
        exit(EXIT_FAILURE);
    }

    EVP_MD_CTX_free(mdctx);
}

// Function to generate random bits following NIST SP 800-90B guidelines
void generate_random_bits(uint8_t *random_bits, int bit_length) {
    uint8_t noise[SOURCE_SIZE];
    collect_noise(noise);
    health_tests(noise);
    double estimated_entropy = estimate_entropy(noise);
    if (estimated_entropy < 2.0) {
        fprintf(stderr, "Insufficient entropy\n");
        exit(EXIT_FAILURE);
    }
    uint8_t conditioned_output[EVP_MAX_MD_SIZE];
    condition(noise, conditioned_output);
    memcpy(random_bits, conditioned_output, bit_length / 8);
}

int main() {
    uint8_t random_bits[32];
    int a = 1;
    while (a <= 10000) {
        generate_random_bits(random_bits, 256);
        printf("Number: %d\n", a);
        for (int i = 0; i < 32; i++) {
            printf("%02x", random_bits[i]);
        }
        printf("\n");
        a++;
    }
    return 0;
}
